open_spiel/python/pytorch/ppo.py

```diff
@@ -0,0 +1,344 @@
+    num_actions, 
+    num_players,
+    player_id=0,
+    num_envs=1,
+    steps_per_batch=128,
+    num_minibatches=4,
+    update_epochs=4,
+    learning_rate=2.5e-4, 
+    num_annealing_updates=None, 
+    gae=True,
+    gamma=0.99,
+    gae_lambda=0.95,
+    normalize_advantages=True,
+    clip_coef=0.2,
+    clip_vloss=True,
+    entropy_coef=0.01,
+    value_coef=0.5,
+    max_grad_norm=0.5,
+    target_kl=None,
+    device='cpu', 
+    writer=None, # Tensorboard SummaryWriter
+    agent_fn=PPOAtariAgent,
+    ):
+    super().__init__()
+
+    self.input_shape = input_shape
+    self.num_actions = num_actions
+    self.num_players = num_players
+    self.player_id = player_id
+    self.device = device
+
+    # Training settings
+    self.num_envs = num_envs
+    self.steps_per_batch = steps_per_batch
+    self.batch_size = self.num_envs * self.steps_per_batch
+    self.num_minibatches = num_minibatches
+    self.minibatch_size = self.batch_size // self.num_minibatches
+    self.update_epochs = update_epochs
+    self.learning_rate = learning_rate
+    self.num_annealing_updates = num_annealing_updates
+
+    # Loss function
+    self.gae = gae
+    self.gamma = gamma
+    self.gae_lambda = gae_lambda
+    self.normalize_advantages = normalize_advantages
+    self.clip_coef = clip_coef
+    self.clip_vloss = clip_vloss
+    self.entropy_coef = entropy_coef
+    self.value_coef = value_coef
+    self.max_grad_norm = max_grad_norm
+    self.target_kl = target_kl
+
+    # Logging
+    self.writer = writer
+
+    # Initialize networks
+    self.network = agent_fn(self.num_actions, self.input_shape, device).to(device)
+    self.optimizer = optim.Adam(self.parameters(), lr=self.learning_rate, eps=1e-5)
+    
+    # Initialize training buffers
+    self.legal_actions_mask = torch.zeros((self.steps_per_batch, self.num_envs, self.num_actions), dtype=torch.bool).to(device)
+    self.obs = torch.zeros((self.steps_per_batch, self.num_envs) + self.input_shape).to(device)
+    self.actions = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.logprobs = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.rewards = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.dones = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+    self.values = torch.zeros((self.steps_per_batch, self.num_envs)).to(device)
+
+    # Initialize counters 
+    self.cur_batch_idx = 0
+    self.total_steps_done = 0
+    self.updates_done = 0
+    self.start_time = time.time()
+
+  def get_value(self, x):
+    return self.network.get_value(x)
+
+  def get_action_and_value(self, x, legal_actions_mask=None, action=None):
+    return self.network.get_action_and_value(x, legal_actions_mask, action)
+
+  def step(self, time_step, is_evaluation=False):
+    if is_evaluation:
+      singular_env = False
+      if not isinstance(time_step, list):
+        time_step = [time_step]
+        singular_env = True
```

Comment:
I would suggest fixing input and output types to always be lists and doing these transformations outside of the algorithm. Changing input/output types may lead to confusing behavior.

@lanctot do you have a preference ?